### PR TITLE
renames Schema\Table to Schema\TableSchema

### LINKED ISSUE #8715: Rename the Schema\Table class
This is a (multiple allowed):
* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.8
* Platform and Target: Apache 2, Ubuntu 14.04, PHP 7.0.5

### What you did

Added a custom schema type to a table.

```
    protected function _initializeSchema(Table $schema)
    {
        $schema->columnType('keywords','json');
        return parent::_initializeSchema($schema);
    }
```

### Expected Behavior

Be able to overload the `_initializeSchema` method without any difficulties.

### Actual Behavior

Have to alias the class name because `Table` is already used.

```
    use Cake\Database\Schema\Table as Schema;

    protected function _initializeSchema(Schema $schema)
    {
        $schema->columnType('keywords','json');
        return parent::_initializeSchema($schema);
    }
```

I would like to have the `Schema\Table` deprecated and a new name given to the class. Seems like the schema and table model are colliding often enough that renaming the class would be a benefit.

### PR DESCRIPTION
**Pull request by thinkingmedia**
fixes #8715
- renamed the class using PhpStorm's refactor feature.
- renamed references in unit tests using strings
- removed use alias from Table

### CODE REVIEW

File: src/Database/Schema/Table.php [hunk 1]
```diff
@@ -14,789 +14,9 @@
...
- * or by incrementally building an instance using
- * methods.
- *
- * Once created Table instances can be added to
- * Schema\Collection objects. They can also be converted into SQL using the
- * createSql(), dropSql() and truncateSql() methods.
+ * @deprecated Provided for backward compatibility. Use TableSchema instead.
  */
-class Table
+class Table extends TableSchema
```
Comment (markstory) [inline src/Database/Schema/Table.php]: You can use class_alias() instead of a subclass.

File: src/Database/Schema/TableSchema.php [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+use Cake\Datasource\ConnectionInterface;
+
+/**
+ * Represents a single table in a database schema.
+ *
+ * Can either be populated using the reflection API's
+ * or by incrementally building an instance using
+ * methods.
+ *
+ * Once created Table instances can be added to
```
Comment (markstory) [inline src/Database/Schema/TableSchema.php]: Need a rename here.

**Comment by thinkingmedia** [comment]
I'll fix this today.

File: src/Database/Schema/Table.php (hunk 1 above)
Comment (thinkingmedia) [inline src/Database/Schema/Table.php]: @markstory does class_alias work with the class loader? Do I have to have a `Table.php` file to ensure backward compatibility.

Comment (thinkingmedia) [inline src/Database/Schema/Table.php]: @markstory I can't get the `@deprecated` annotation to work with class_alias, and we would want IDEs to flag the class as deprecated. PhpStorm will show a slash across the class name as an example. So I prefer to keep it this way for now. If people switch to TableSchema there won't be any performance issues for them.

**Comment by thinkingmedia** [comment]
done.

File: src/Database/Schema/Table.php (hunk 1 above)
Comment (chinpei215) [inline src/Database/Schema/Table.php]: @thinkingmedia  It is not a performance issue. @markstory says that your change breaks existing applications. The following code would throw a fatal error.
```php
namespace App\Model\Table;

use Cake\ORM\Table;
use Cake\Database\Schema\Table as Schema

class Articles extends Table
{
    protected function _initializeSchema(Schema $schema) { // Fatal error, as TableSchema is not a subclass of Schema
    
    }
}

```

Comment (thinkingmedia) [inline src/Database/Schema/Table.php]: @chinpei215 your example is wrong. Schema is an alias for Table and Table is a subclass of TableSchema. No error will be thrown.

Comment (thinkingmedia) [inline src/Database/Schema/Table.php]: image

As you can see it works as I wanted. The class is indicated as deprecated on old code but does not break anything.

Comment (chinpei215) [inline src/Database/Schema/Table.php]: @thinkingmedia  Didn't you change `Cake\ORM\Table`? Then wouldn't Cake use the deprecated class in the core?

Comment (thinkingmedia) [inline src/Database/Schema/Table.php]: @chinpei215 this PR changes all the code to use TableSchema.

Anyone updating to this PR will see all usages of `Cake\Database\Schema\Table` as depreciated without breaking.

Comment (chinpei215) [inline src/Database/Schema/Table.php]: @thinkingmedia Indeed you did change `ORM\Cake\Table`. Then does the line 17 in your capture really works? It seems to change the signature of parent method.
```php
// parent signature
protected function _initializeSchema(\Cake\Database\Schema\TableSchema $schema);
// child signature
protected function _initializeSchema(\Cake\Database\Schema\Table $schema);
```
Even if it works, the following code wouldn't work:
```php
use Cake\Database\Schema\Table as Schema;

$schema = $articles->schema(); // \Cake\Database\Schema\TableSchema
if (!$schema instanceof Schema) { // \Cake\Database\Schema\Table
    throw new \Exception('Never come here');
} 
```

Comment (thinkingmedia) [inline src/Database/Schema/Table.php]: @chinpei215 ah that is a very good catch. Nice example. Thanks.

Comment (chinpei215) [inline src/Database/Schema/Table.php]: @thinkingmedia You are welcome :)

Comment (markstory) [inline src/Database/Schema/Table.php]: @chinpei215 Explained what I was getting at far better than I did. Thank you.

File: src/ORM/Table.php [hunk 3]
```diff
@@ -460,17 +460,17 @@ public function schema($schema = null)
...
      * ```
      *
-     * @param \Cake\Database\Schema\Table $table The table definition fetched from database.
-     * @return \Cake\Database\Schema\Table the altered schema
+     * @param \Cake\Database\Schema\TableSchema $table The table definition fetched from database.
+     * @return \Cake\Database\Schema\TableSchema the altered schema
      * @api
      */
-    protected function _initializeSchema(Schema $table)
+    protected function _initializeSchema(TableSchema $table)
```
Comment (markstory) [inline src/ORM/Table.php]: These type hint changes might cause strict errors in applications. I'll check that tonight.

File: src/Database/Schema/Table.php (hunk 1 above)
Comment (chinpei215) [inline src/Database/Schema/Table.php]: @markstory No problem :)

File: src/ORM/Table.php (hunk 3 above)
Comment (markstory) [inline src/ORM/Table.php]: I checked, and these typehint changes will cause strict errors for existing applications:

```
Strict (2048): Declaration of App\Model\Table\BookmarksTable::_initializeSchema() should be compatible with Cake\ORM\Table::_initializeSchema(Cake\Database\Schema\TableSchema $table) [APP/Model/Table/BookmarksTable.php, line 14]
```

I think reverting the type hint changes might be the best option if you still want to land this in `master`.

Comment (thinkingmedia) [inline src/ORM/Table.php]: I think the point was to fix type hint so that we didn't have to use an alias. Let me think about this for a bit.

Comment (markstory) [inline src/ORM/Table.php]: It was, but at the same time we can't break people's applications when they upgrade.

Comment (thinkingmedia) [inline src/ORM/Table.php]: @markstory do you think class_alias will fix this problem?

Comment (markstory) [inline src/ORM/Table.php]: Only if the typehints aren't changed in `Cake\ORM\Table`. Changing any of the typehints in commonly extended classes will cause strict errors in existing code.

Comment (chinpei215) [inline src/ORM/Table.php]: @markstory I didn't know that :stuck_out_tongue:
Then #8612 and #8709 might cause strict errors too, if people are using overridden methods having effected typehints in their applications. It would be better to document it when we release them.

Comment (markstory) [inline src/ORM/Table.php]: I think #8612 is safe as no type hints changed. #8709 could cause issues, but I suspect fewer people have typehints against the HTTP/Client internals than override _initializeSchema()

**Comment by ADmad** [comment]
Having to alias `Cake\Database\Schema\Table` was the primary reason the class was renamed. But since the typehint can't be changed without causing problem the issues stands as it is, so I don't see a point in renaming the class.

**Comment by dereuromark** [comment]
So this should target 3.next and could then work with class_alias()?

**Comment by thinkingmedia** [comment]
lol, this PR didn't turn out the way I was hoping.

My only recommendation is to either

- create a new `_initializeSchema` function with a new name and mark the old one deprecated
- abdomen this PR and just leave things as is.

**Comment by ADmad** [comment]
Why don't we just have a callback `buildSchema()` like we have `buildValidator()` ?

**Comment by lorenzo** [comment]
@ADmad The callback is called `_initializeSchema()`

**Comment by ADmad** [comment]
I meant using the event system :)

File: src/ORM/Table.php (hunk 3 above)
Comment (lorenzo) [inline src/ORM/Table.php]: Yeah, we need to revert this change and maybe add another function with the new typehint

**Comment by markstory** [comment]
Closing as this can no longer be merged, and in its current state would break backwards compatibility.
